✨ Start drafting a proposal for combining reported state

MikeSpreitzer opened the pull request:
<!--
Thanks for creating a pull request!

If this is your first time, please make sure to review CONTRIBUTING.MD.

Please copy the appropriate `:text:` or icon to the beginning of your PR title:

:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization

-->
## Summary
This is one proposal for an API for addressing #2017. See the changes in `types.go`, and the new doc file (which should be folded into the user guide, which should exist).

## Related issue(s)

ezrasilvera commented:
This should be included and be done as part of the overall design of 2017.  
As we agreed, for all such features, we first review and agree on the requirements, we then create a design proposal (@vMaroon is doing that for this  2017) , and finally we have an implementation PR. 
@MikeSpreitzer You are welcomed to add comments and proposals to the design doc @vMaroon created and we can/should discuss it there.  
There is no point in splitting the design into two proposals.

effi-ofer commented on docs/content/direct/combined-status.md:
```diff
@@ -0,0 +1,148 @@
+
+1. The combined reported state appears in a new kind of object, one per (workload object, BindingPolicy object) pair.
+
+1. A user can request a list without aggregation, possibly after filtering, but certainly with a limit on list length. The expectation is that such a list makes sense only if the length of the list will be modest. For users that want access to the full reported state from each WEC for a large number of WECs, KubeStellar should have an abstraction that gives the users access --- in a functional wa ...
```
Is the mailbox still a concept now that we replaced IMBS with ITS?

MikeSpreitzer replied:
We still have mailbox namespaces in the ITS. Changing the term from IMBS to ITS does not change that. Also, it does not change the fact that the mailbox namespaces are an internal implementation detail.

MikeSpreitzer commented:
@ezrasilvera: take this in the spirit of a design document that includes some Go code. See how there is markdown here too?

vMaroon commented:
@MikeSpreitzer I think @ezrasilvera's point is that I had already started working on this issue, and have asked to work on it in a live meeting before the bug-fixing break, and again after ezra made the features doc.

Are we going to both work on this in parallel and have the team vote? Are we going to merge into a single solution? Do you reject the proposal I made or did you make this independently? 

I think we should sync.

MikeSpreitzer commented:
@vMaroon: I assumed that it was OK for other people to also submit design proposals. As you can see, I have read your design proposal and made some comments on it. I think that we are in conversation. This proposal addresses comments that I made on the other one and adds a few ideas of its own. I think that the clearest way to show those ideas is with an API and examples.

MikeSpreitzer commented:
I have marked this PR as a "draft", trying to make the intent clearer. This is meant as a part of a design discussion, showing ideas of what the API might look like.

MikeSpreitzer commented:
Apparently GitHub 'draft' status means "not for review". That seems not quite right, I am asking people to consider the ideas here. I will mark it as "hold" because I am not asking this to be merged now.

/hold

vMaroon commented:
@MikeSpreitzer matching SQL SELECT functionality is smart and is a capability I wanted to achieve in iterations, going for it directly is sensible and is a very powerful capability that can cover pretty much any use-case or user requirement.
However, I believe the current API might be overly complex and too narrowly focused on direct SQL analogies—such as the use of terms like 'rows' and 'columns'. Simplifying this API to make it more intuitive and less reliant on SQL expertise would be beneficial.

The flexibility I wanted to achieve in my proposal can be started as groundwork provided by a non-invasive extension to your proposal:
`StatusReturn` could be streamlined as suggested, then repurposed into a standalone API object that users can define and reference by name. Additionally, we would permit the direct inlining of this API within BindingPolicies. This dual capability would satisfy both simple and complex use-cases.
This way, status collection intents can be declared once and reused in different BindingPolicies, for different object selectors, while specific intents can be inlined.

By enabling the latter, we allow an environment where common-use templates can be provided out-of-the-box, developed by the community and developed by users in their specific uses. This also opens up the path for future higher-level APIs that could apply status collection intents to multiple BindingPolicies, as I tried to initiate in the StatusPolicy proposal.

I think it makes sense to proceed with this "third" proposal, focusing initially on developing a user-friendly API for StatusReturn. What do you think?

pdettori commented on docs/content/direct/combined-status.md:
```diff
@@ -0,0 +1,148 @@
+1. The SELECT statement can have a WHERE clause that filters out some of the rows.
+
+1. The SELECT statmement either does aggregation or does not. In the case of not doing aggregation, the SELECT statement simply has a collection of named expressions defining the columns of its output.
+
```
s/statmement/statement/
